### ISSUE TITLE
[new rule] Require that TypeScript tests use a typed jest.mock factory

### ISSUE THREAD

**Issue by NotWoods** [jest-community/eslint-plugin-jest#1313]
A new rule that requires type-checking. 

`jest.mock` and `jest.doMock` take an optional generic, representing the return type of the factory. For example,
```ts
// The optional type argument provides typings for the module factory
jest.mock<typeof import('../moduleName')>('../moduleName', () => {
  return jest.fn(() => 42);
});
```

Without specifying `typeof import('../moduleName')`, Jest's types doesn't know what the module looks like and will allow anything to be returned in the factory.

I'd like to have an ESLint rule to enforce that the generic parameter is used whenever a factory function is used. Jest auto-mocks (no factory) and virtual mocks (no corresponding actual module) don't need to have the generic parameter.

Examples of **incorrect** code for this rule:
```ts
jest.mock('../moduleName', () => {
  return jest.fn(() => 42);
});
jest.mock('./module', () => ({
  ...jest.requireActual('./module'),
  foo: jest.fn()
}));
jest.mock('random-num', () => {
  return jest.fn(() => 42);
});
```

Examples of **correct** code for this rule:
```ts
// Uses typeof import()
jest.mock<typeof import('../moduleName')>('../moduleName', () => {
  return jest.fn(() => 42);
});
jest.mock<typeof import('./module')>('./module', () => ({
  ...jest.requireActual('./module'),
  foo: jest.fn()
}));

// Uses custom type
jest.mock<() => number>('random-num', () => {
  return jest.fn(() => 42);
});
// No factory
jest.mock('random-num');
// Virtual mock
jest.mock(
  '../moduleName',
  () => {
    return jest.fn(() => 42)
  },
  {virtual: true},
);
```

**Auto-fixable:** Yes

**Comment by G-Rath** [comment]
This sounds ok, and won't require type-checking, just `@typescript-eslint/parser` as generics are syntax so they're represented as AST nodes.
